aspnet/fundamentals/hosting.rst

```diff
@@ -0,0 +1,296 @@
+
+.. code-block:: c#
+  :emphasize-lines: 4-8
+
+  // configure host, in Main()
+  var host = new WebHostBuilder()
+            .UseKestrel()
+            .ConfigureServices(s => {
+                s.AddSingleton<IFormatter, LowercaseFormatter>();
+            })
+            .ConfigureLogging(f => f.AddConsole(LogLevel.Debug))
+            .UseStartup<Startup>()
+            .Build();
+
+  host.Run();
+
+Create the ``IFormatter`` interface and an implementation:
+
+.. code-block:: c#
+
+  public interface IFormatter
+  {
+    string Format(string input);
+  }
+
+  public class LowercaseFormatter : IFormatter
+  {
+    public string Format(string input)
+      {
+        return input.ToLower();
+      }
+  }
+
+Specify its dependencies in the ``Startup`` class's constructor:
+
+.. code-block:: none
+    :emphasize-lines: 6,24
+
+    public class Startup
+    {
+        private readonly ILogger _logger;
+        private readonly IFormatter _formatter;
+        
+        public Startup(ILoggerFactory loggerFactory, IFormatter formatter)
+        {
+            _logger = loggerFactory.CreateLogger<Startup>();
+            _formatter = formatter;
+        }
+        
+        public void ConfigureServices(IServiceCollection services)
+        {
+            _logger.LogDebug($"Total Services Initially: {services.Count}");
+
+            // register additional services
+
+            _logger.LogDebug($"Total Services Afterward: {services.Count}");
+        }
+
+        public void Configure(IApplicationBuilder app, IFormatter formatter)
+        {
+            _logger.LogDebug("Configure() started...");
+            app.Run(async (context) => await context.Response.WriteAsync(formatter.Format("Hi!")));
+            _logger.LogDebug("Configure() complete.");
+        }
+    }
+
+Both the constructor and the ``Configure`` method can request the ``IFormatter`` service (and in this case, since it is configured as a Singleton, both of them will get the same instance of the service). Within the ``Configure`` method above, both ``formatter`` and ``_formatter`` are available to use.
+
+.. note:: The ``Startup`` class supports dependency injection in its constructor and the ``Configure`` method, but not the ``ConfigureServices`` method. However, ``ConfigureServices`` can access any registered service through its ``IServiceCollection`` parameter, if necessary.
+
+Ordering Importance
+^^^^^^^^^^^^^^^^^^^
+
+The order in which the host is configured using the ``WebHostBuilder`` is important. The host defaults to using predefined environment variables first, if defined. These environment variables must use the format ``ASPNETCORE_{configurationKey}``, so for example to set the URLs the server will listen on by default, you would set ``ASPNETCORE_URLS``.
```

Comment:
Need to replace "The order in which the host is configured using the ``WebHostBuilder`` is important." with something like
WebHostBuilder config setting are read in the following order
A 
B 
C
The last one wins